File: Python/2019/3/main.py
import re

D = {
    'U': (0, -1),
    'D': (0, 1),
    'L': (-1, 0),
    'R': (1, 0),
}

with open("input.txt") as f:
    moves = [[(x, int(y)) for x, y in re.findall(r"([UDRL])(\d+)", line)] for line in f.readlines()]

pos = [(0, 0), (0, 0)]
C = set()
I = set()

d = 0

for dir, dist in moves[0]:
    for _ in range(dist):
        pos[0] = (pos[0][0] + D[dir][0], pos[0][1] + D[dir][1])
        d += 1
        C.add((pos[0], d))

d = 0

seen = set([x[0] for x in C])

for dir, dist in moves[1]:
    for _ in range(dist):
        pos[1] = (pos[1][0] + D[dir][0], pos[1][1] + D[dir][1])
        d += 1
        if pos[1] in seen:
            I.add((pos[1], d))

S = []

for coord, steps in I:
    s = [x[1] for x in C if x[0] == coord][0]
    S.append(steps + s)

closest = min(I, key=lambda k: abs(k[0][0]) + abs(k[0][1]))
print("Part 1:", abs(closest[0][0]) + abs(closest[0][1]))
print("Part 2:", min(S))

File: Go/2019/3/main.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"

	"github.com/TomWatson6/AdventOfCode/Go/2019/utils"
)

type Coord struct {
	X int
	Y int
}

func (c *Coord) Add(co Coord) {
	c.X += co.X
	c.Y += co.Y
}

func (c Coord) Distance() int {
	return int(math.Abs(float64(c.X)) + math.Abs(float64(c.Y)))
}

func main() {
	lines, err := utils.ReadLines("input.txt")
	if err != nil {
		panic(err)
	}

	w1, err := getWire(lines[0])
	if err != nil {
		panic(err)
	}

	w2, err := getWire(lines[1])
	if err != nil {
		panic(err)
	}

	inter := intersections(w1, w2)

	smallest := int(1e9)

	for _, i := range inter {
		val := i.Distance()
		if val < smallest {
			smallest = val
		}
	}

	fmt.Printf("Part 1: %d\n", smallest)

	s1, err := getSteps(lines[0], inter)
	if err != nil {
		panic(err)
	}

	s2, err := getSteps(lines[1], inter)
	if err != nil {
		panic(err)
	}

	smallest = int(1e9)

	for c, s := range s1 {
		total := s + s2[c]
		if total < smallest {
			smallest = total
		}
	}

	fmt.Printf("Part 2: %d\n", smallest)
}

func getSteps(line string, inter []Coord) (map[Coord]int, error) {
	stepsMap := map[Coord]int{}
	coordMap := map[Coord]bool{}

	for _, c := range inter {
		if c.X == 0 && c.Y == 0 {
			continue
		}

		coordMap[c] = true
	}

	pos := Coord{0, 0}
	steps := 0

	for _, val := range strings.Split(line, ",") {
		dir := val[0]
		length, err := strconv.Atoi(val[1:])
		if err != nil {
			return nil, err
		}

		d := getDirection(rune(dir))

		for range length {
			pos.Add(d)
			steps += 1
			if _, ok := coordMap[pos]; ok {
				stepsMap[pos] = steps
			}
		}
	}

	return stepsMap, nil
}

func getWire(line string) (map[Coord]bool, error) {
	wire := map[Coord]bool{
		{0, 0}: true,
	}
	pos := Coord{0, 0}

	for _, val := range strings.Split(line, ",") {
		dir := val[0]
		length, err := strconv.Atoi(val[1:])
		if err != nil {
			return nil, err
		}

		d := getDirection(rune(dir))

		for range length {
			pos.Add(d)
			wire[pos] = true
		}
	}

	return wire, nil
}

func getDirection(dir rune) Coord {
	switch dir {
	case 'U':
		return Coord{-1, 0}
	case 'D':
		return Coord{1, 0}
	case 'L':
		return Coord{0, -1}
	case 'R':
		return Coord{0, 1}
	}

	return Coord{}
}

func intersections(w1, w2 map[Coord]bool) []Coord {
	intersections := []Coord{}

	for w := range w1 {
		if w.X == 0 && w.Y == 0 {
			continue
		}

		if _, ok := w2[w]; ok {
			intersections = append(intersections, w)
		}
	}

	return intersections
}
